### ISSUE TITLE
Services do not update on render

### ISSUE THREAD

**Issue by alettieri** [aws-amplify/amplify-ui#4170]
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have tried disabling all browser extensions or using a different browser
- [X] I have tried deleting the node_modules folder and reinstalling my dependencies
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Authenticator

### How is your app built?

next.js

### Please describe your bug.

We run some custom registration logic during the signup process and override the `handleSignUp` service in the `Authenticator` component. We manage the account registration on a separate `/registration` path. 

It looks like the Authenticator component initializes services on the initial render, but doesn't respect the passed-in services on consecutive renders.  It would be great if the `services` object would update on every re-render so that our custom service runs.

What seems to happen is that our `handleSignUp` override function is ignored and instead the `default` handler is called.

### What's the expected behaviour?

I expect the `onSignUp` handler to get called when we submit a new registration form, it doesn't get called.

### Help us reproduce the bug!

It's fairly easy for me to replicate this issue, by navigating from our `login` page back to our `registration` page. I then fill in the form and click the submit button.

### Code Snippet

Here is an example of our registration page, I have removed some code to try and keep the code a little more terse and easy to follow:

```typescript
const Registration: AppPage<{ defaultEmail?: string }> = (props) => {
   /... removed code for brevity
    return (
        <>
            <Head>
                <title key="title">Telemetry : Registration</title>
            </Head>
            <RegistrationForm
                defaultEmail={props.defaultEmail}
                onSignUp={registerUser.mutateAsync}
            />
        </>
    );
};

export default Registration;
```

The `registerUser.mutateAsync` query mutation object, we leverage `react-query` for API mutations.

We have some custom registration logic we need to run as part of our sign up process:

```typescript

    const registerUser = useMutation<
        ISignUpResult,
        Error,
        IRegistrationFormData
    >(
        async ({ username, password }) => {
            const params: SignUpParams = {
                username,
                password,
                attributes: {
                    email: username,
                },
                autoSignIn: {
                    enabled: true, // we try to sign users in automatically and use an email invitation link to help achieve this.
                },
            };

            const signUpResult = await Auth.signUp(params);

            const registrationInvite: IRegistration = {
                email: signUpResult.user.getUsername(),
                username: signUpResult.userSub,
                billing_plan: billingPlan,
                utm_campaign: utmCampaign,
                utm_content: utmContent,
                utm_medium: utmMedium,
                utm_source: utmSource,
                utm_term: utmTerm,
            };


             await apiClient.createRegistration(
                  registrationInvite
              );

            return signUpResult; // Important for the Authenticator `handleSignUp` service to process the result
        },
    );
```
    
    
    Finally, here is what the `RegistrationForm` component looks like, it is essentially where the `Authenticator` component is rendered:
    
```typescript
    const RegistrationForm = ({
    defaultEmail,
    onSignUp,
}: IRegistrationFormProps) => {
    const { apiClientWithoutBearerToken } = React.useContext(ApiContext);
    const { route, toSignUp } = useAuthenticator((c) => [c.route]);

    const formFields = React.useMemo<AuthenticatorProps['formFields']>(() => {
        return {
            signUp: {
                username: {
                    label: 'Email*',
                    isRequired: true,
                    placeholder: 'Enter your email address',
                    defaultValue: defaultEmail,
                },
                password: {
                    label: 'Password*',
                    isRequired: true,
                    placeholder: 'Enter your password',
                },
                confirm_password: {
                    label: 'Confirm Password*',
                    isRequired: true,
                    placeholder: 'Please confirm your password',
                },
            },
        };
    }, [defaultEmail]);

    const validateEmail = React.useCallback(
        async (email: string) => {
           // ... code removed for brevity
            
        },
        [apiClientWithoutBearerToken]
    );

    const validatePreferredUsername = React.useCallback(
        async (
            formData: Pick<IValidationFormData, 'username'>,
            touchData: Pick<IValidationTouchData, 'username'>
        ) => {
           // ... code removed for brevity
        },
        [validateEmail]
    );

    React.useEffect(() => {
        if (route === 'signIn') {
            /** This is just to ensure we never accidentally switch to sign in on the registration page */
            toSignUp();
        }
    }, [toSignUp, route]);

    /**
     * @note: The tabs are hidden in a global css file
     */
    return (
        <Authenticator
            initialState="signUp"
            className="login-background"
            components={components}
            formFields={formFields}
            services={{
                validateFormPassword,
                validatePreferredUsername,
                handleSignUp: onSignUp, // passed in `onSignUp` prop
            }}
            socialProviders={['google']}
        />
    );
};
```

**Comment by alettieri** [comment]
Note, if I pull in the `initializeMachine` function and re-initialize the state machine, the services run as expected. 

Minor update to the `RegistrationForm` component:

```typescript
const { route, toSignUp, initializeMachine } = useAuthenticator((c) => [
        c.route,
    ]);
```

```typescript
    React.useEffect(() => {
        initializeMachine({
            initialState: 'signUp',
            components,
            formFields,
            services: {
                validateFormPassword,
                validatePreferredUsername,
                handleSignUp: onSignUp,
            },
        });
    }, [initializeMachine, validatePreferredUsername, onSignUp, formFields]);
```

**Comment by alettieri** [comment]
I'm starting to understand what the issue is. Our app is actually wrapped in an authenticator provider  `Authenticator.Provider`, and it looks like the `setup` process won't run after the first successful initialization. 

https://github.com/aws-amplify/amplify-ui/blob/4b46399ead7435e71f074f6c152e72132aa23a91/packages/react-core/src/Authenticator/hooks/useAuthenticatorInitMachine/useAuthenticatorInitMachine.tsx#L19-L22

Meaning the state does not get initialized when the application switches from one route to another. 

It may be that we'll need to break out when and where the auth provider is rendered in addition to having `Authenticator` as a child component further down the context.

**Comment by alettieri** [comment]
Okay, I think I solved my own issue here. It seems like it is related to some related issues with the way the provider is setup. 

We render the authenticator component in multiple pages, some are protected and some unprotected. What we essentially did was break out the registration UI into it's own path and render the authenticator component there. 

Before we wrapped the application pages in one `Authenticator.Provider` component which managed the machine state for the entire application. The issue is that the initialization process lives in the `Authenticator` component, so once the route changed from `setup` to say `signIn` or `signUp` the state was maintained in the Provider, meaning if we unmount, then remount the `Authenticator` component, it never had a chance to re-initialize because the route value never reached the `setup` stage. 

To solve this, I actually broke out when and where we render the `Authenticator.Provider`. We handle this conditionally based on wether the page is protected or not protected. 

```typescript
const renderApp = (
    ui: React.ReactNode,
    {
        isPageProtected = true,
    }: {
        isPageProtected: boolean;
    } = {
        isPageProtected: true,
     }
) => {
    if (!isPageProtected) {
        return <Authenticator.Provider>{ui}</Authenticator.Provider>;
    }

    return (
        <AppErrorBoundary>
            <Authenticator.Provider>
                <Session>
                   {ui}
                </Session>
            </Authenticator.Provider>
        </AppErrorBoundary>
    );
};
```

> note: The `Session` component contains the `Authenticator` component.
